### PR TITLE
[common-utils] - Bookworm - enabled nonFreePackages option for installing manpages-posix, manpages-posix-dev packages

### PR DESCRIPTION
**Pull request by gauravsaini04**
Enabled non-free option in `sources.list.d/debian.sources` file for downloading and installation of `manpages-posix` & `manpages-posix-dev` packages to address [#760](https://github.com/devcontainers/images/issues/760)

Test with cmd: `devcontainer features test . -f common-utils --skip-autogenerated --filter "bookworm"  --preserve-test-containers`

### CODE REVIEW

File: src/common-utils/main.sh [hunk 1]
```diff
@@ -104,18 +104,22 @@ install_debian_packages() {
...
-        sed -i "s/deb http:\/\/deb\.debian\.org\/debian ${VERSION_CODENAME}-backports main/deb http:\/\/deb\.debian\.org\/debian ${VERSION_CODENAME}-backports main contrib non-free/" /etc/apt/sources.list
-        sed -i "s/deb-src http:\/\/deb\.debian\.org\/debian ${VERSION_CODENAME}-backports main/deb http:\/\/deb\.debian\.org\/debian ${VERSION_CODENAME}-backports main contrib non-free/" /etc/apt/sources.list
-        # Handle bullseye location for security https://www.debian.org/releases/bullseye/amd64/release-notes/ch-information.en.html
-        sed -i "s/deb http:\/\/security\.debian\.org\/debian-security ${VERSION_CODENAME}-security main/deb http:\/\/security\.debian\.org\/debian-security ${VERSION_CODENAME}-security main contrib non-free/" /etc/apt/sources.list
-        sed -i "s/deb-src http:\/\/security\.debian\.org\/debian-security ${VERSION_CODENAME}-security main/deb http:\/\/security\.debian\.org\/debian-security ${VERSION_CODENAME}-security main contrib non-free/" /etc/apt/sources.list
+        if [ ! -e "/etc/apt/sources.list" ]; then
```
Comment (samruddhikhandale) [inline src/common-utils/main.sh:107]: Can you show me the contents of `/etc/apt/sources.list` file before we do string replacement?

Comment (samruddhikhandale) [reply src/common-utils/main.sh:107]: Also, output of `ls /etc/apt`

Comment (gauravsaini04) [reply src/common-utils/main.sh:107]: Tested for `bullseye` with disabled `nonFreePackages` :
```
root ➜ / $ cat etc/apt/sources.list
# deb http://snapshot.debian.org/archive/debian/20240130T194419Z bullseye main
deb http://deb.debian.org/debian bullseye main
# deb http://snapshot.debian.org/archive/debian-security/20240130T194419Z bullseye-security main
deb http://deb.debian.org/debian-security bullseye-security main
# deb http://snapshot.debian.org/archive/debian/20240130T194419Z bullseye-updates main
deb http://deb.debian.org/debian bullseye-updates main
```

Tested for `bullseye` again after enabling `nonFreePackages` :
```
# deb http://snapshot.debian.org/archive/debian/20240130T194419Z bullseye main
deb http://deb.debian.org/debian bullseye main contrib non-free
# deb http://snapshot.debian.org/archive/debian-security/20240130T194419Z bullseye-security main
deb http://deb.debian.org/debian-security bullseye-security main
# deb http://snapshot.debian.org/archive/debian/20240130T194419Z bullseye-updates main
deb http://deb.debian.org/debian bullseye-updates main contrib non-free
```

Comment (gauravsaini04) [reply src/common-utils/main.sh:107]: output of ls -la /etc/apt/sources.list.d :

root ➜ / $ ls -la /etc/apt/sources.list.d
total 16
drwxr-xr-x 1 root root 4096 Feb  7 00:29 .
drwxr-xr-x 1 root root 4096 Jan 30 19:44 ..
-rw-r--r-- 1 root root  477 Feb  7 00:29 debian.sources

Comment (gauravsaini04) [reply src/common-utils/main.sh:107]: Contents of /etc/apt/sources.list.d/debian.sources before we do string replacement :
```
Types: deb
# http://snapshot.debian.org/archive/debian/20240130T194419Z
URIs: http://deb.debian.org/debian
Suites: bookworm bookworm-updates
Components: main
Signed-By: /usr/share/keyrings/debian-archive-keyring.gpg

Types: deb
# http://snapshot.debian.org/archive/debian-security/20240130T194419Z
URIs: http://deb.debian.org/debian-security
Suites: bookworm-security
Components: main
Signed-By: /usr/share/keyrings/debian-archive-keyring.gpg
```

After replacing strings globally :
```
Types: deb
# http://snapshot.debian.org/archive/debian/20240130T194419Z
URIs: http://deb.debian.org/debian
Suites: bookworm bookworm-updates
components: main contrib non-free
Signed-By: /usr/share/keyrings/debian-archive-keyring.gpg
 
Types: deb
# http://snapshot.debian.org/archive/debian-security/20240130T194419Z
URIs: http://deb.debian.org/debian-security
Suites: bookworm-security
components: main contrib non-free
Signed-By: /usr/share/keyrings/debian-archive-keyring.gpg
```

File: src/common-utils/main.sh [hunk 2]
```diff
@@ -104,18 +104,22 @@ install_debian_packages() {
...
-        # Handle bullseye location for security https://www.debian.org/releases/bullseye/amd64/release-notes/ch-information.en.html
-        sed -i "s/deb http:\/\/security\.debian\.org\/debian-security ${VERSION_CODENAME}-security main/deb http:\/\/security\.debian\.org\/debian-security ${VERSION_CODENAME}-security main contrib non-free/" /etc/apt/sources.list
-        sed -i "s/deb-src http:\/\/security\.debian\.org\/debian-security ${VERSION_CODENAME}-security main/deb http:\/\/security\.debian\.org\/debian-security ${VERSION_CODENAME}-security main contrib non-free/" /etc/apt/sources.list
+        if [ ! -e "/etc/apt/sources.list" ]; then 
+            sed -i -E "s/Components: main/components: main contrib non-free/" /etc/apt/sources.list.d/debian.sources
+        else
+            # Bring in variables from /etc/os-release like VERSION_CODENAME
+            sed -i -E "s/deb http:\/\/(deb|httpredir)\.debian\.org\/debian ${VERSION_CODENAME} main/deb http:\/\/\1\.debian\.org\/debian ${VERSION_CODENAME} main contrib non-free/" /etc/apt/sources.list
```
[the remark this answers is not in the document]
Comment (gauravsaini04) [reply src/common-utils/main.sh:111]: ## when replacing when Suites have values: bookworm bookworm-updates and not when its value is: bookworm-security
```
root ➜ / $ sed '/Suites: bookworm bookworm-updates/{N;s/Components: main/Components: main contrib non-free/}' /etc/apt/sources.list.d/debian.sources

Types: deb
# http://snapshot.debian.org/archive/debian/20240130T194419Z
URIs: http://deb.debian.org/debian
Suites: bookworm bookworm-updates
Components: main contrib non-free
Signed-By: /usr/share/keyrings/debian-archive-keyring.gpg

Types: deb
# http://snapshot.debian.org/archive/debian-security/20240130T194419Z
URIs: http://deb.debian.org/debian-security
Suites: bookworm-security
Components: main
Signed-By: /usr/share/keyrings/debian-archive-keyring.gpg

root ➜ / $ apt-get update -y
Get:1 http://deb.debian.org/debian bookworm InRelease [151 kB]
Get:2 http://deb.debian.org/debian bookworm-updates InRelease [52.1 kB]
Get:3 http://deb.debian.org/debian-security bookworm-security InRelease [48.0 kB]
Get:4 http://deb.debian.org/debian bookworm/main amd64 Packages [8787 kB]
Get:5 http://deb.debian.org/debian bookworm-updates/main amd64 Packages [12.7 kB]
Get:6 http://deb.debian.org/debian-security bookworm-security/main amd64 Packages [137 kB]
Fetched 9188 kB in 2s (5832 kB/s)                         
Reading package lists... Done
root ➜ / $ apt-get install -y -no-install-recommends manpages-posix manpages-posix-dev
E: Command line option 'n' [from -no-install-recommends] is not understood in combination with the other options.
root ➜ / $ apt-get install -y --no-install-recommends manpages-posix manpages-posix-dev
Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
E: Unable to locate package manpages-posix
E: Unable to locate package manpages-posix-dev
```

## when replacing only when Suites have value bookworm-security and not when the value is bookworm bookworm-updates,  then also packages not getting installed

```
root ➜ / $ sed '/Suites: bookworm-security/{N;s/Components: main/Components: main contrib non-free/}' /etc/apt/sources.list.d/debian.sources

Types: deb
# http://snapshot.debian.org/archive/debian/20240130T194419Z
URIs: http://deb.debian.org/debian
Suites: bookworm bookworm-updates
Components: main
Signed-By: /usr/share/keyrings/debian-archive-keyring.gpg

Types: deb
# http://snapshot.debian.org/archive/debian-security/20240130T194419Z
URIs: http://deb.debian.org/debian-security
Suites: bookworm-security
Components: main contrib non-free
Signed-By: /usr/share/keyrings/debian-archive-keyring.gpg
root ➜ / $ apt-get update -y
Get:1 http://deb.debian.org/debian bookworm InRelease [151 kB]
Get:2 http://deb.debian.org/debian bookworm-updates InRelease [52.1 kB]
Get:3 http://deb.debian.org/debian-security bookworm-security InRelease [48.0 kB]
Get:4 http://deb.debian.org/debian bookworm/main amd64 Packages [8787 kB]
Get:5 http://deb.debian.org/debian bookworm-updates/main amd64 Packages [12.7 kB]
Get:6 http://deb.debian.org/debian-security bookworm-security/main amd64 Packages [137 kB]
Fetched 9188 kB in 2s (5832 kB/s)                         
Reading package lists... Done
root ➜ / $ apt-get install -y -no-install-recommends manpages-posix manpages-posix-dev
E: Command line option 'n' [from -no-install-recommends] is not understood in combination with the other options.
root ➜ / $ apt-get install -y --no-install-recommends manpages-posix manpages-posix-dev
Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
E: Unable to locate package manpages-posix
E: Unable to locate package manpages-posix-dev
```

So, we have to keep using "main contrib non-free" for Suites as bookworm bookworm-updates and bookworm-security as well for the installation of the manpages-posix and manpages-posix-dev to be successful in the container

[the remark this answers is not in the document]
Comment (samruddhikhandale) [reply src/common-utils/main.sh:111]: I was able to install `manpages-posix` and `manpages-posix-dev` in `bookworm`. I had to add components `non-free-firmware non-free` to http://deb.debian.org/debian

image

[the remark this answers is not in the document]
Comment (gauravsaini04) [reply src/common-utils/main.sh:111]: done

[the remark this answers is not in the document]
Comment (gauravsaini04) [reply src/common-utils/main.sh:111]: Tests added

File: src/common-utils/main.sh (hunk 1 above)
Comment (samruddhikhandale) [inline src/common-utils/main.sh:107]: Can you also add another condition to validate that `/etc/apt/sources.list.d/debian.sources` file exists?

File: src/common-utils/main.sh [hunk 3]
```diff
@@ -104,18 +104,22 @@ install_debian_packages() {
...
-        sed -i "s/deb-src http:\/\/deb\.debian\.org\/debian ${VERSION_CODENAME}-backports main/deb http:\/\/deb\.debian\.org\/debian ${VERSION_CODENAME}-backports main contrib non-free/" /etc/apt/sources.list
-        # Handle bullseye location for security https://www.debian.org/releases/bullseye/amd64/release-notes/ch-information.en.html
-        sed -i "s/deb http:\/\/security\.debian\.org\/debian-security ${VERSION_CODENAME}-security main/deb http:\/\/security\.debian\.org\/debian-security ${VERSION_CODENAME}-security main contrib non-free/" /etc/apt/sources.list
-        sed -i "s/deb-src http:\/\/security\.debian\.org\/debian-security ${VERSION_CODENAME}-security main/deb http:\/\/security\.debian\.org\/debian-security ${VERSION_CODENAME}-security main contrib non-free/" /etc/apt/sources.list
+        if [ ! -e "/etc/apt/sources.list" ]; then 
+            sed -i '/Suites: bookworm bookworm-updates/{N;s/Components: main/Components: main non-free non-free-firmware/}' /etc/apt/sources.list.d/debian.sources
```
Comment (samruddhikhandale) [inline src/common-utils/main.sh:108]: Can you only apply `sed` when the URI is `http://deb.debian.org/debian` ? because that's the only change needed.

**Review by samruddhikhandale — CHANGES_REQUESTED** [review]

File: src/common-utils/main.sh (hunk 3 above)
Comment (gauravsaini04) [reply src/common-utils/main.sh]: done 👍

File: src/common-utils/main.sh [hunk 4]
```diff
@@ -104,18 +104,22 @@ install_debian_packages() {
 
     # Needed for adding manpages-posix and manpages-posix-dev which are non-free packages in Debian
```
Comment (samruddhikhandale) [inline src/common-utils/main.sh:105]: Can we update patch version of this Feature? https://github.com/devcontainers/features/blob/e5054dba652b09902ec43aa45bf05386a28a341c/src/common-utils/devcontainer-feature.json#L3

**Review by samruddhikhandale — CHANGES_REQUESTED** [review]

File: src/common-utils/main.sh (hunk 4 above)
Comment (gauravsaini04) [reply src/common-utils/main.sh:105]: done 👍

**Review by samruddhikhandale — APPROVED** [review]
Looks good ⚡

**Pull request merged**